codegenerator/cli/templates/dynamic/codegen/src/Types.res.hbs

```diff
@@ -202,6 +202,174 @@ type eventLog<'a> = {
+    let hasRegistration: t<'eventArgs> => bool
+  } = {
+    type loaderReturn
+
+    type t<'eventArgs> = {
+      contractName: string,
+      eventName: string,
+      topic0: string,
+      mutable loaderHandler: option<loaderHandler<'eventArgs, loaderReturn>>,
+      mutable contractRegister: option<contractRegister<'eventArgs>>,
+      mutable eventOptions: option<eventOptions>,
+    }
+
+    let getLoaderHandler = (t: t<'eventArgs>): option<loaderHandler<'eventArgs, 'loaderReturn>> =>
+      t.loaderHandler->(
+        Utils.magic: option<loaderHandler<'eventArgs, loaderReturn>> => option<
+          loaderHandler<'eventArgs, 'loaderReturn>,
+        >
+      )
+
+    let getContractRegister = (t: t<'eventArgs>): option<contractRegister<'eventArgs>> =>
+      t.contractRegister
+
+    let getEventOptions = ({eventOptions, topic0}: t<'eventArgs>): eventOptions =>
+      switch eventOptions {
+      | Some(eventOptions) => eventOptions
+      | None => getDefaultEventOptions(~topic0)
+      }
+
+    let hasRegistration = ({loaderHandler, contractRegister}) =>
+      loaderHandler->Belt.Option.isSome || contractRegister->Belt.Option.isSome
+
+    let make = (~topic0, ~contractName, ~eventName) => {
+      contractName,
+      eventName,
+      topic0,
+      loaderHandler: None,
+      contractRegister: None,
+      eventOptions: None,
+    }
+
+    type eventNamespace = {contractName: string, eventName: string}
+    exception DuplicateEventRegistration(eventNamespace)
+
+    let setEventOptions = (t: t<'eventArgs>, value: eventOptions, ~logger=Logging.logger) => {
+      switch t.eventOptions {
+      | None => t.eventOptions = Some(value)
+      | Some(_) =>
+        let eventNamespace = {contractName: t.contractName, eventName: t.eventName}
+        DuplicateEventRegistration(eventNamespace)->ErrorHandling.mkLogAndRaise(
+          ~logger=Logging.createChildFrom(~logger, ~params=eventNamespace),
+          ~msg="Duplicate eventOptions in handlers not allowed",
+        )
+      }
+    }
+
+    let setLoaderHandler = (
+      t: t<'eventArgs>,
+      value: loaderHandler<'eventArgs, 'loaderReturn>,
+      ~eventOptions,
+      ~logger=Logging.logger,
+    ) => {
+      switch t.loaderHandler {
+      | None =>
+        t.loaderHandler =
+          value
+          ->(Utils.magic: loaderHandler<'eventArgs, 'loaderReturn> => loaderHandler<
+            'eventArgs,
+            loaderReturn,
+          >)
+          ->Some
+      | Some(_) =>
+        let eventNamespace = {contractName: t.contractName, eventName: t.eventName}
+        DuplicateEventRegistration(eventNamespace)->ErrorHandling.mkLogAndRaise(
+          ~logger=Logging.createChildFrom(~logger, ~params=eventNamespace),
+          ~msg="Duplicate registration of event handlers not allowed",
+        )
+      }
+
+      switch eventOptions {
+      | Some(eventOptions) => t->setEventOptions(eventOptions, ~logger)
+      | None => ()
+      }
+    }
```

Comment:
This is a new abstraction for registering handlers with with options on an event module